test: Migrate CI to GitHub

evansims opened the pull request:
### Changes

This pull request replaces our current CircleCI continuous integration suite for auth0-react with a new workflow using GitHub Actions.

### References

Please see internal ticket SDK-4457.

### Testing

This pull request replaces our continuous integration suite with one built in GitHub's native Actions ecosystem. It aligns with other migrations taking place across our repositories.

### Checklist

* [x] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
* [x] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)
* [x] All existing and new tests complete without errors
* [x] Rubocop passes on all added/modified files
* [x] All active GitHub checks have passed

stevehobbsdev commented on .github/workflows/publish.yml:
```diff
@@ -0,0 +1,41 @@
+          mkdir -p $HOME/.gem
+          touch $HOME/.gem/credentials
+          chmod 0600 $HOME/.gem/credentials
+          printf -- "---\n:rubygems_api_key: ${GEM_HOST_API_KEY}\n" > $HOME/.gem/credentials
+          gem build *.gemspec
+          gem push *.gem
+        env:
+          GEM_HOST_API_KEY: "${{secrets.RUBYGEMS_AUTH_TOKEN}}"
```
Is there a need to both create the `credentials` file _and_ create an environment variable? Wouldn't we just use the env var?

evansims replied:
Not familiar enough with Ruby to make an authoritative statement one way or the other on that, but this is implemented per [GitHub's suggested approach](https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-ruby#publishing-gems). My assumption was the environment variable was being used elsewhere in the `gem` process.

evansims commented:
@stevehobbsdev Thanks for your review! I've removed the `gem-release` dependency and updated the publishing workflow. I think I caught all the related bits and pieces, but let me know if I happened to overlook one.

Did a bit of research and found the reason GitHub suggests this approach. RubyGems < 3.0.5 did not support `GEM_HOST_API_KEY` variable for `gem push`, and instead required the use of that credentials file. So this approach covers all cases. I don't expect we'll have such an old version of it floating about these days, so it's indeed safe to remove that aspect 👍
